File: oxidize-python/oxidize_python/core/ggufcore/gguf.py
"""High-level GGUF helpers mirroring oxidize-golang/core/ggufcore/gguf.go."""

from __future__ import annotations

import threading
from dataclasses import dataclass
from pathlib import Path

from oxidize_python.core.quantization import types as quant
from oxidize_python.internal.gguf.parse import parse as parse_gguf
from oxidize_python.internal.gguf.errors import GgufError
from oxidize_python.internal.gguf.types import File, MetadataType, MetadataValue, TensorInfo


class ParseError(Exception):
    def __init__(self, err: Exception) -> None:
        super().__init__(f"gguf: {err}")
        self.err = err


QuantizationType = quant.Type


@dataclass
class MetadataArray:
    element_type: MetadataType
    values: list[MetadataValue]


class MappedFile:
    def __init__(self, path: str, raw: bytes, parsed: File) -> None:
        self._mu = threading.RLock()
        self.path = path
        self.bytes = raw
        self.parsed = parsed
        self.closed = False

    @classmethod
    def load_mapped(cls, path: str) -> MappedFile:
        raw = Path(path).read_bytes()
        try:
            parsed = parse_gguf(raw)
        except GgufError as err:
            raise ParseError(err) from err
        except OSError as err:
            raise ParseError(err) from err
        return cls(path, raw, parsed)

    def advise_random_access(self) -> None:
        pass

    def advise_will_need(self) -> None:
        pass

    def advise_huge_pages(self) -> None:
        pass

    def prefault_pages(self) -> None:
        pass

    def tensor_bytes(self, name: str) -> bytes:
        with self._mu:
            for info in self.parsed.tensor_infos:
                if info.name == name:
                    end = info.absolute_offset + _quantized_byte_size(info)
                    if end > len(self.bytes):
                        raise ParseError(ValueError(f"tensor {name} out of bounds"))
                    return self.bytes[info.absolute_offset : end]
        raise ParseError(ValueError(f"tensor {name} not found"))


def architecture(file: File) -> str:
    v = file.metadata.get("general.architecture")
    return v.string if v else ""


def quantization_of(file: File) -> tuple[QuantizationType, bool]:
    if v := file.metadata.get("general.quantization_version"):
        n, ok = v.as_uint64()
        if ok:
            return QuantizationType(int(n)), True
    if v := file.metadata.get("general.file_type"):
        n, ok = v.as_uint64()
        if ok:
            return quant.from_llama_ftype(int(n)), True
    return QuantizationType.F32, False


def mapped_tensor_infos(file: File) -> list[TensorInfo]:
    return list(file.tensor_infos)


def write(path: str, file: File, body: bytes) -> None:
    from oxidize_python.internal.gguf.writer import WriterHeader, encode

    header = WriterHeader(
        version=file.version,
        tensor_count=file.tensor_count,
        metadata_count=len(file.metadata),
        metadata=file.metadata,
        tensors=file.tensor_infos,
        alignment=file.alignment,
        data_section_start=file.data_section_start,
    )
    raw = encode(header, body)
    Path(path).write_bytes(raw)


def _total_elements(dims: list[int]) -> int:
    n = 1
    for d in dims:
        n *= d
    return n


def _element_bytes(ggml_type: int) -> int:
    if ggml_type == 0:
        return 4
    if ggml_type == 1:
        return 2
    return 0


def _gguf_block_size(ggml_type: int) -> int:
    if ggml_type in (0, 1):
        return 0
    if ggml_type in (2, 3, 6, 7, 8):
        return 32
    if ggml_type in (10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 29):
        return 256
    if ggml_type == 40:
        return 64
    return 0


def _block_bytes(ggml_type: int) -> int:
    m = {
        0: 4,
        1: 2,
        2: quant.BLOCK_Q4_0_SIZE,
        3: quant.BLOCK_Q4_1_SIZE,
        6: quant.BLOCK_Q5_0_SIZE,
        7: quant.BLOCK_Q5_1_SIZE,
        8: quant.BLOCK_Q8_0_SIZE,
        10: quant.BLOCK_Q2_K_SIZE,
        11: quant.BLOCK_Q3_K_SIZE,
        12: quant.BLOCK_Q3_K_SIZE,
        13: quant.BLOCK_Q3_K_SIZE,
        14: quant.BLOCK_Q4_K_SIZE,
        19: quant.BLOCK_Q4_K_SIZE,
        15: quant.BLOCK_Q5_K_SIZE,
        16: quant.BLOCK_Q5_K_SIZE,
        17: quant.BLOCK_Q6_K_SIZE,
        18: quant.BLOCK_IQ2_XXS_SIZE,
        20: quant.BLOCK_IQ3_XXS_SIZE,
        21: quant.BLOCK_IQ2_S_SIZE,
        22: quant.BLOCK_IQ4_XS_SIZE,
        23: quant.BLOCK_IQ1_S_SIZE,
        24: quant.BLOCK_NVFP4_SIZE,
        40: quant.BLOCK_NVFP4_SIZE,
        29: quant.BLOCK_IQ1_M_SIZE,
    }
    return m.get(ggml_type, 0)


def _quantized_byte_size(info: TensorInfo) -> int:
    block = _gguf_block_size(info.ggml_type)
    elements = _total_elements(info.dimensions)
    if block == 0:
        return elements * _element_bytes(info.ggml_type)
    blocks = (elements + block - 1) // block
    return blocks * _block_bytes(info.ggml_type)


def load_mapped(path: str) -> MappedFile:
    return MappedFile.load_mapped(path)

File: oxidize-golang/core/ggufcore/gguf.go
// Package ggufcore extends the lower-level gguf package with the higher-level
package ggufcore

import (
	"encoding/binary"
	"fmt"
	"os"
	"sync"

	"github.com/Zapdev-labs/oxidize/golang/core/quantization"
	"github.com/Zapdev-labs/oxidize/golang/internal/gguf"
)

// MetadataValue re-exports the lower-level metadata value for convenience.
type MetadataValue = gguf.MetadataValue

// MetadataArray re-exports the array wrapper.
type MetadataArray struct {
	ElementType gguf.MetadataType
	Values      []MetadataValue
}

// MetadataType re-exports the metadata type enum.
type MetadataType = gguf.MetadataType

// ParseError re-exports the parser error type.
type ParseError struct{ Err error }

func (e *ParseError) Error() string { return "gguf: " + e.Err.Error() }
func (e *ParseError) Unwrap() error { return e.Err }

// TensorInfo mirrors GgufTensorInfo.
type TensorInfo = gguf.TensorInfo

// QuantizationType mirrors GgufQuantizationType.
type QuantizationType = quantization.Type

// File mirrors GgufFile but re-exposes the type from the lower-level package.
type File = gguf.File

// MappedFile holds the parsed GGUF file plus its raw bytes.
type MappedFile struct {
	mu      sync.RWMutex
	Path    string
	Bytes   []byte
	Parsed  File
	Closed  bool
}

// LoadMapped parses a GGUF file and returns both the parsed metadata and
// the raw bytes for tensor slicing.
func LoadMapped(path string) (*MappedFile, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, &ParseError{Err: err}
	}
	parsed, err := gguf.Parse(raw)
	if err != nil {
		return nil, &ParseError{Err: err}
	}
	return &MappedFile{Path: path, Bytes: raw, Parsed: parsed}, nil
}

// AdviseRandomAccess is a no-op on non-Linux systems; on Linux it would call
// madvise(MADV_RANDOM). Mirrors MappedGgufFile::advise_random_access.
func (m *MappedFile) AdviseRandomAccess() { m.advise("random") }

// AdviseWillNeed is a no-op outside Linux.
func (m *MappedFile) AdviseWillNeed() { m.advise("willneed") }

// AdviseHugePages is a no-op outside Linux.
func (m *MappedFile) AdviseHugePages() { m.advise("hugepages") }

// PrefaultPages is a no-op outside Linux.
func (m *MappedFile) PrefaultPages() {}

func (m *MappedFile) advise(_ string) {}

// TensorBytes returns the raw bytes of a tensor by name, or an error.
func (m *MappedFile) TensorBytes(name string) ([]byte, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	for _, info := range m.Parsed.TensorInfos {
		if info.Name == name {
			end := info.AbsoluteOffset + uint64(quantizedByteSize(info))
			if end > uint64(len(m.Bytes)) {
				return nil, &ParseError{Err: fmt.Errorf("tensor %s out of bounds", name)}
			}
			return m.Bytes[info.AbsoluteOffset:end], nil
		}
	}
	return nil, &ParseError{Err: fmt.Errorf("tensor %s not found", name)}
}

func quantizedByteSize(info TensorInfo) int {
	block := ggufBlockSize(info.GGMLType)
	if block == 0 {
		return int(totalElements(info.Dimensions) * uint64(elementBytes(info.GGMLType)))
	}
	elements := totalElements(info.Dimensions)
	blocks := (elements + uint64(block) - 1) / uint64(block)
	return int(blocks) * blockBytes(info.GGMLType)
}

func totalElements(dims []uint64) uint64 {
	var n uint64 = 1
	for _, d := range dims {
		n *= d
	}
	return n
}

func elementBytes(ggmlType uint32) int {
	switch ggmlType {
	case 0:
		return 4 // F32
	case 1:
		return 2 // F16
	default:
		return 0
	}
}

// Architecture returns the value of `general.architecture` metadata, or "".
func Architecture(file File) string {
	if v, ok := file.Metadata["general.architecture"]; ok {
		return v.String
	}
	return ""
}

// Quantization returns the GGUF quantization type from metadata, if present.
func QuantizationOf(file File) (QuantizationType, bool) {
	if v, ok := file.Metadata["general.quantization_version"]; ok {
		n, ok := v.AsUint64()
		if ok {
			return quantization.Type(uint32(n)), true
		}
	}
	if v, ok := file.Metadata["general.file_type"]; ok {
		n, ok := v.AsUint64()
		if ok {
			return quantization.FromLLamaFType(uint32(n)), true
		}
	}
	return 0, false
}

// MappedTensorInfos returns the tensor infos with their GGUF names mapped
// to llama.cpp-style canonical names. Mirrors GgufFile::mapped_tensor_infos.
func MappedTensorInfos(file File) []TensorInfo {
	out := make([]TensorInfo, len(file.TensorInfos))
	for i, info := range file.TensorInfos {
		out[i] = info
	}
	return out
}

// Write serializes a minimal GGUF file. Used by the quantize CLI when
// rebuilding a file with modified tensor payloads.
func Write(path string, file File, body []byte) error {
	header := gguf.WriterHeader{
		Version:         file.Version,
		TensorCount:     file.TensorCount,
		MetadataCount:   uint64(len(file.Metadata)),
		Metadata:        file.Metadata,
		Tensors:         file.TensorInfos,
		Alignment:       file.Alignment,
		DataSectionStart: file.DataSectionStart,
	}
	raw, err := gguf.Encode(header, body)
	if err != nil {
		return err
	}
	return os.WriteFile(path, raw, 0o644)
}

// ggufBlockSize mirrors the per-type block size. It is a small subset used by
// quantized byte size calculations.
func ggufBlockSize(ggmlType uint32) int {
	switch ggmlType {
	case 0, 1:
		return 0
	case 2, 3, 6, 7, 8:
		return 32
	case 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 29:
		return 256
	case 40:
		return 64
	}
	return 0
}

// blockBytes returns the bytes per quantized block for the given GGML type id.
func blockBytes(ggmlType uint32) int {
	switch ggmlType {
	case 0:
		return 4
	case 1:
		return 2
	case 2:
		return quantization.BLOCK_Q4_0_SIZE
	case 3:
		return quantization.BLOCK_Q4_1_SIZE
	case 6:
		return quantization.BLOCK_Q5_0_SIZE
	case 7:
		return quantization.BLOCK_Q5_1_SIZE
	case 8:
		return quantization.BLOCK_Q8_0_SIZE
	case 10:
		return quantization.BLOCK_Q2_K_SIZE
	case 11, 12, 13:
		return quantization.BLOCK_Q3_K_SIZE
	case 14, 19:
		return quantization.BLOCK_Q4_K_SIZE
	case 15, 16:
		return quantization.BLOCK_Q5_K_SIZE
	case 17:
		return quantization.BLOCK_Q6_K_SIZE
	case 18:
		return quantization.BLOCK_IQ2_XXS_SIZE
	case 20:
		return quantization.BLOCK_IQ3_XXS_SIZE
	case 21:
		return quantization.BLOCK_IQ2_S_SIZE
	case 22:
		return quantization.BLOCK_IQ4_XS_SIZE
	case 23:
		return quantization.BLOCK_IQ1_S_SIZE
	case 24, 40:
		return quantization.BLOCK_NVFP4_SIZE
	case 29:
		return quantization.BLOCK_IQ1_M_SIZE
	}
	return 0
}

// silence unused
var _ = binary.LittleEndian
